src/main/drivers/flash.c

```diff
@@ -30,92 +30,118 @@
 
 #include "flash.h"
 #include "flash_m25p16.h"
+#include "flash_w25n01g.h"
 
 #include "common/time.h"
 
 #include "drivers/bus_spi.h"
 #include "drivers/io.h"
 #include "drivers/time.h"
 
-static flashPartitionTable_t flashPartitionTable;
-static int flashPartitions = 0;
-
+static flashDriver_t flashDrivers[] = {
 
 #ifdef USE_SPI
-static bool flashSpiInit(void)
-{
+
 #ifdef USE_FLASH_M25P16
-    return m25p16_init(0);
+    {
+        .init = m25p16_init,
+        .isReady = m25p16_isReady,
+        .waitForReady = m25p16_waitForReady,
+        .eraseSector = m25p16_eraseSector,
+        .eraseCompletely = m25p16_eraseCompletely,
+        .pageProgram = m25p16_pageProgram,
+        .readBytes = m25p16_readBytes,
+        .getGeometry = m25p16_getGeometry,
+        .flush = NULL,
+        .detected = true
+    },
+#endif
+
+#ifdef USE_FLASH_W25N01G
+    {
+        .init = w25n01g_init,
+        .isReady = w25n01g_isReady,
+        .waitForReady = w25n01g_waitForReady,
+        .eraseSector = w25n01g_eraseSector,
+        .eraseCompletely = w25n01g_eraseCompletely,
+        .pageProgram = w25n01g_pageProgram,
+        .readBytes = w25n01g_readBytes,
+        .getGeometry = w25n01g_getGeometry,
+        .flush = w25n01g_flush,
+        .detected = true
+    },
 #endif
-    return false;
-}
-#endif // USE_SPI
 
-bool flashDeviceInit(void)
-{
-#ifdef USE_SPI
-    return flashSpiInit();
 #endif
 
-    return false;
+};
+
+static flashDriver_t *flash;
+
+static bool flashDeviceInit(void)
+{
+    for (uint32_t idx = 0; idx <= sizeof(flashDrivers) / sizeof(flashDrivers[0]); idx++)
+    {
+        flash->detected = flashDrivers[idx].init(0);
+        if (flash->detected)
+        {
+            flash = &flashDrivers[idx];
+            break;
+        }
+    }
+    return flash->detected;
 }
 
 bool flashIsReady(void)
 {
-#ifdef USE_FLASH_M25P16
-    return m25p16_isReady();
-#endif
-    return false;
+    // prevent the machine cycle from crashing if there is no external flash memory
+    if (!flash->detected) {
```

Comment:
This is not a perfectly safe approach to test for absence of flash, as this requires flashDeviceInit to have initialized the `flash` structure pointer before this check can be used. 
It also fails if flashfs (USE_FLASHFS) and none of the flash drivers are configured.
Easiest solution to this is to remove the `detected` field from the flash structure and put this as a static variable into the module flash.c and initialized with `false`. 
Added benefit of removing `detected` is that the flash structs array for the drivers could go to read-only memory (not RAM) if const is added to array definition as the function pointers do not change during run-time.

Reply:
An alternative and probably the easiest solution is to use as test `if (flash == NULL)` as the flash variable only gets assigned if a driver was successfully initialized.